fl4health/losses/deep_mmd_loss.py

```diff
@@ -0,0 +1,224 @@
+                -Dyy_org / sigma
+            )
+            Kxy = (1 - epsilon) * torch.exp(-((Dxy / sigma0) ** L) - Dxy_org / sigma) + epsilon * torch.exp(
+                -Dxy_org / sigma
+            )
+        else:
+            Kx = torch.exp(-Dxx / sigma0)
+            Ky = torch.exp(-Dyy / sigma0)
+            Kxy = torch.exp(-Dxy / sigma0)
+
+        return self.h1_mean_var_gram(Kx, Ky, Kxy, is_var_computed, use_1sample_U)
+
+    def train_kernel(self, X: torch.Tensor, Y: torch.Tensor) -> None:
+        """Train the kernel."""
+
+        self.featurizer.train()
+        self.sigmaOPT.requires_grad = True
+        self.sigma0OPT.requires_grad = True
+        self.epsilonOPT.requires_grad = True
+
+        features = torch.cat([X, Y], 0)
+
+        # ------------------------------
+        #  Train deep network for MMD-D
+        # ------------------------------
+        # Initialize optimizer
+        self.optimizer_F.zero_grad()
+        # Compute output of deep network
+        model_output = self.featurizer(features)
+        # Compute epsilon, sigma and sigma_0
+        ep = torch.exp(self.epsilonOPT) / (1 + torch.exp(self.epsilonOPT))
+        sigma = self.sigmaOPT**2
+        sigma0_u = self.sigma0OPT**2
+        # Compute Compute J (STAT_u)
+        mmd_value_temp, mmd_var_temp, _ = self.MMDu(
+            Fea=model_output,
+            len_s=X.shape[0],
+            Fea_org=features.view(features.shape[0], -1),
+            sigma=sigma,
+            sigma0=sigma0_u,
+            epsilon=ep,
+        )
+        if mmd_var_temp is None:
+            raise AssertionError("Error: Variance of MMD is not computed. Please set is_var_computed=True.")
+        mmd_std_temp = torch.sqrt(mmd_var_temp + 10 ** (-8))
+        STAT_u = torch.div(-1 * mmd_value_temp, mmd_std_temp)
+        # Compute gradient
+        STAT_u.backward()
+        # Update weights using gradient descent
+        self.optimizer_F.step()
+        return
+
+    def compute_kernel(self, X: torch.Tensor, Y: torch.Tensor) -> torch.Tensor:
+        """Train the kernel."""
+
+        self.featurizer.eval()
+        self.sigmaOPT.requires_grad = False
+        self.sigma0OPT.requires_grad = False
+        self.epsilonOPT.requires_grad = False
+
+        features = torch.cat([X, Y], 0)
+
+        # Compute output of deep network
+        model_output = self.featurizer(features)
+        # Compute epsilon, sigma and sigma_0
+        ep = torch.exp(self.epsilonOPT) / (1 + torch.exp(self.epsilonOPT))
+        sigma = self.sigmaOPT**2
+        sigma0_u = self.sigma0OPT**2
+        # Compute Compute J (STAT_u)
+        mmd_value_temp, _, _ = self.MMDu(
+            Fea=model_output,
+            len_s=X.shape[0],
+            Fea_org=features.view(features.shape[0], -1),
+            sigma=sigma,
+            sigma0=sigma0_u,
+            epsilon=ep,
+        )
+
+        return mmd_value_temp
+
+    def forward(self, Xs: torch.Tensor, Xt: torch.Tensor) -> torch.Tensor:
```

Comment:
So based on our implementation, am I correct in thinking that during training, we'll also be updating the kernel values at each step based on the batches? If this doesn't quite work, another thing we could do is update the deep kernels periodically, like we do with MKMMD where we "optimize" the kernels for a frozen set of models.

Reply:
Yes exactly that's what we are doing. We first update kernel for each batch then compute value.